tests/Microsoft.Identity.Web.Test/TokenAcquisitionAuthorityTests.cs

```diff
@@ -309,5 +309,109 @@ public void MergeExtraQueryParameters_MergedOptionsNull_Test()
             // Assert
             Assert.Null(mergedDict);
         }
+
+        [Theory]
+        [InlineData("https://localhost:1234")]
+        [InlineData("")]
+        [InlineData(null)]
+        public void ManagedIdCacheKey_Test(string? clientId)
+        {
+            // Arrange
+            string defaultKey = "SYSTEM";
+            ManagedIdentityOptions managedIdentityOptions = new()
+            {
+                UserAssignedClientId = clientId
+            };
+
+            // Act
+            string key = TokenAcquisition.GetCacheKeyForManagedId(managedIdentityOptions);
+
+            // Assert
+            if (string.IsNullOrEmpty(clientId))
+            {
+                Assert.Equal(defaultKey, key);
+            }
+            else
+            {
+                Assert.Equal(clientId, key);
+            }
+        }
+
+        [Theory]
+        [InlineData("https://localhost:1234")]
+        [InlineData("")]
+        [InlineData(null)]
+        public async void GetOrBuildManagedIdentity_TestAsync(string? clientId)
+        {
+            // Arrange
+            ManagedIdentityOptions managedIdentityOptions = new()
+            {
+                UserAssignedClientId = clientId
+            };
+            MergedOptions mergedOptions = new();
+            BuildTheRequiredServices();
+            InitializeTokenAcquisitionObjects();
+
+            // Act
+            IManagedIdentityApplication app1 = 
+                await _tokenAcquisition.GetOrBuildManagedIdentityApplication(mergedOptions, managedIdentityOptions);
+            IManagedIdentityApplication app2 = 
+                await _tokenAcquisition.GetOrBuildManagedIdentityApplication(mergedOptions, managedIdentityOptions);
+
+            // Assert
+            Assert.Same(app1, app2);
+        }
+
+        [Theory]
+        [InlineData("https://localhost:1234")]
+        [InlineData(null)]
+        public async void GetOrBuildManagedIdentity_TestConcurrencyAsync(string? clientId)
+        {
+            // Arrange
+            int numThreads = 20;
```

Comment:
> 20 [](http://example.com/codeflow?start=28&length=3)

reason for 20? just curious.

Reply:
I think avoiding magic numbers is good; this value could either be defined as a constant at the top of the file, or at least a comment on why 20 is used.

Reply:
I agree I should mark it const but I don't want to put it at the top of the file vs. the top of the method because no other method uses it and then it is separated by 300+ lines from where it is used.

As far as 20 being a magic number, it is, but I don't know of any metric that could be used to determine an optimal number of threads for testing concurrency. Did you have any thoughts on it @westin-m?

Reply:
> I don't want to put it at the top of the file vs. the top of the method

That's probably fine, I just say file because it's more conventional. Not too important.

> determine an optimal number of threads for testing concurrency

I can't say a ton about this, but I've seen others do three tests for concurrency: one thread (for correct behavior); _number of typical processors_ +  1  threads (depends on customers' production envs); and something large like 1000 threads to increase context switching and contention. You could swap 20 out for 33 (32 cores on a reasonable server + 1) to have a little more rhyme and reason, but feel free to find a second opinion.

Reply:
Keegan recommended using ThreadPool.MaxThreads so I did that :)